Missing channel announcement

While working on PR #1493 I believe I discovered a bug related to missing announcements to connected peer.

Problem can be see [here](https://travis-ci.org/lightningnetwork/lnd/jobs/399082974) 
The integration tests failed with an error "channel not opened before timeout".

This error is not consistent and has a probability for showing up (my guess is 25%). 

Let me explain the situation using 3 nodes:Alice, Bob, Carol

1. Alice (P2Pport 19555) connects with Bob (19558). They open a channel between them.

2. Carol (19561) connects with Alice.

3. Bob restarts. At startup both Alice and Bob tries to restore the connection. Lets assume that Bob managed to create the connection, from port 59756 to 19555. Alice detected that she does not need to continue trying to set a connection with Bob.

Alice log:
```
2018-07-02 11:28:57.038 [DBG] CMGR: Attempting to connect to 0397268f5962e3cdf0fe3d9f8acac4b2384f74a8ace109ed400e19727b4cc9da6f@127.0.0.1:19558 (reqid 13)
2018-07-02 11:28:57.039 [DBG] CMGR: Failed to connect to 0397268f5962e3cdf0fe3d9f8acac4b2384f74a8ace109ed400e19727b4cc9da6f@127.0.0.1:19558 (reqid 13): dial tcp 127.0.0.1:19558: connect: connection refused
2018-07-02 11:28:57.039 [DBG] CMGR: Retrying connection to 0397268f5962e3cdf0fe3d9f8acac4b2384f74a8ace109ed400e19727b4cc9da6f@127.0.0.1:19558 (reqid 13) in 5s
2018-07-02 11:28:57.539 [INF] SRVR: New inbound connection from 127.0.0.1:59756
2018-07-02 11:28:57.539 [INF] SRVR: finalizing connection to 0397268f5962e3cdf0fe3d9f8acac4b2384f74a8ace109ed400e19727b4cc9da6f, inbound=true
2018-07-02 11:28:57.539 [DBG] PEER: Sending Init to 127.0.0.1:59756
2018-07-02 11:28:57.539 [DBG] CMGR: Canceling: 0397268f5962e3cdf0fe3d9f8acac4b2384f74a8ace109ed400e19727b4cc9da6f@127.0.0.1:19558 (reqid 13)
2018-07-02 11:28:57.540 [DBG] PEER: Received Init from 127.0.0.1:59756  
```

4. Now, Carol opens a channel with Alice. All good so Alice tries to broadcast the channel announcement to Bob

Alice log:
```

2018-07-02 11:29:01.680 [INF] DISC: Broadcasting batch of 4 new announcements
2018-07-02 11:29:01.680 [DBG] SRVR: Broadcasting 1 messages
2018-07-02 11:29:01.680 [DBG] SRVR: Broadcasting 1 messages
2018-07-02 11:29:01.680 [DBG] SRVR: Broadcasting 1 messages
2018-07-02 11:29:01.680 [DBG] SRVR: Broadcasting 1 messages
2018-07-02 11:29:02.039 [DBG] CMGR: Attempting to connect to 0397268f5962e3cdf0fe3d9f8acac4b2384f74a8ace109ed400e19727b4cc9da6f@127.0.0.1:19558 (reqid 13)
2018-07-02 11:29:02.042 [DBG] CMGR: Ignoring connection for canceled connreq=0397268f5962e3cdf0fe3d9f8acac4b2384f74a8ace109ed400e19727b4cc9da6f@127.0.0.1:19558 (reqid 13)

```

Result - Bob is not aware of the channel between Carol and Alice. 

This is exactly the situation in the integration test:

- Alice and Bob connected.
- testSwitchCircuitPersistence - restarts Alice and Bob. They will restore the connection by not always it will be by Alice
- In case connection restored by Bob, testSwitchOfflineDelivery will fail.

@halseth I managed to reproduce it. Please see PR #1730.

The PR basically changes Bob's P2P port and restart Bob's node while Bob has a channel with Alice. When Bob starts Alice can't connect it (since she has its old port number) and Bob is the one to restore the connection. 

You can see the impact by running integration tests on your development machine. 

You can also see the build https://travis-ci.org/lightningnetwork/lnd/builds/416489274 which shows 2 failures. The second  https://travis-ci.org/lightningnetwork/lnd/jobs/416489277 is what we are looking for:

>  --- FAIL: TestLightningNetworkDaemon/switch_circuit_persistence (29.10s)
>     	lnd_test.go:74: Failed: (switch circuit persistence): exited with error: 
>     		*errors.errorString Bob(0): timeout waiting for channel(2e5b51f05a965f7580f6636fa1a2c08c24ca5142a45ead19ed97bd6ba8c13558:0) open: channel not opened before timeout

--- a/lntest/node.go
+++ b/lntest/node.go
@@ -204,6 +204,16 @@ type HarnessNode struct {
 
 	chanWatchRequests chan *chanWatchRequest
 
+	// For each outpoint, we'll track an integer which denotes the number of
+	// edges seen for that channel within the network. When this number
+	// reaches 2, then it means that both edge advertisements has propagated
+	// through the network.
+	openChans   map[wire.OutPoint]int
+	openClients map[wire.OutPoint][]chan struct{}
+
+	closedChans  map[wire.OutPoint]struct{}
+	closeClients map[wire.OutPoint][]chan struct{}
+
 	quit chan struct{}
 	wg   sync.WaitGroup
 
@@ -242,6 +252,11 @@ func newNode(cfg nodeConfig) (*HarnessNode, error) {
 		cfg:               &cfg,
 		NodeID:            nodeNum,
 		chanWatchRequests: make(chan *chanWatchRequest),
+		openChans:         make(map[wire.OutPoint]int),
+		openClients:       make(map[wire.OutPoint][]chan struct{}),
+
+		closedChans:  make(map[wire.OutPoint]struct{}),
+		closeClients: make(map[wire.OutPoint][]chan struct{}),
 	}, nil
 }
 
@@ -673,16 +688,6 @@ func (hn *HarnessNode) lightningNetworkWatcher() {
 		}
 	}()
 
-	// For each outpoint, we'll track an integer which denotes the number
-	// of edges seen for that channel within the network. When this number
-	// reaches 2, then it means that both edge advertisements has
-	// propagated through the network.
-	openChans := make(map[wire.OutPoint]int)
-	openClients := make(map[wire.OutPoint][]chan struct{})
-
-	closedChans := make(map[wire.OutPoint]struct{})
-	closeClients := make(map[wire.OutPoint][]chan struct{})
-
 	for {
 		select {
 
@@ -699,21 +704,21 @@ func (hn *HarnessNode) lightningNetworkWatcher() {
 					Hash:  *txid,
 					Index: newChan.ChanPoint.OutputIndex,
 				}
-				openChans[op]++
+				hn.openChans[op]++
 
 				// For this new channel, if the number of edges
 				// seen is less than two, then the channel
 				// hasn't been fully announced yet.
-				if numEdges := openChans[op]; numEdges < 2 {
+				if numEdges := hn.openChans[op]; numEdges < 2 {
 					continue
 				}
 
 				// Otherwise, we'll notify all the registered
 				// clients and remove the dispatched clients.
-				for _, eventChan := range openClients[op] {
+				for _, eventChan := range hn.openClients[op] {
 					close(eventChan)
 				}
-				delete(openClients, op)
+				delete(hn.openClients, op)
 			}
 
 			// For each channel closed, we'll mark that we've
@@ -726,14 +731,14 @@ func (hn *HarnessNode) lightningNetworkWatcher() {
 					Hash:  *txid,
 					Index: closedChan.ChanPoint.OutputIndex,
 				}
-				closedChans[op] = struct{}{}
+				hn.closedChans[op] = struct{}{}
 
 				// As the channel has been closed, we'll notify
 				// all register clients.
-				for _, eventChan := range closeClients[op] {
+				for _, eventChan := range hn.closeClients[op] {
 					close(eventChan)
 				}
-				delete(closeClients, op)
+				delete(hn.closeClients, op)
 			}
 
 		// A new watch request, has just arrived. We'll either be able
@@ -748,30 +753,34 @@ func (hn *HarnessNode) lightningNetworkWatcher() {
 				// If this is an open request, then it can be
 				// dispatched if the number of edges seen for
 				// the channel is at least two.
-				if numEdges := openChans[targetChan]; numEdges >= 2 {
+				if numEdges := hn.openChans[targetChan]; numEdges >= 2 {
 					close(watchRequest.eventChan)
 					continue
 				}
 
 				// Otherwise, we'll add this to the list of
 				// watch open clients for this out point.
-				openClients[targetChan] = append(openClients[targetChan],
-					watchRequest.eventChan)
+				hn.openClients[targetChan] = append(
+					hn.openClients[targetChan],
+					watchRequest.eventChan,
+				)
 				continue
 			}
 
 			// If this is a close request, then it can be
 			// immediately dispatched if we've already seen a
 			// channel closure for this channel.
-			if _, ok := closedChans[targetChan]; ok {
+			if _, ok := hn.closedChans[targetChan]; ok {
 				close(watchRequest.eventChan)
 				continue
 			}
 
 			// Otherwise, we'll add this to the list of close watch
 			// clients for this out point.
-			closeClients[targetChan] = append(closeClients[targetChan],
-				watchRequest.eventChan)
+			hn.closeClients[targetChan] = append(
+				hn.closeClients[targetChan],
+				watchRequest.eventChan,
+			)
 
 		case <-hn.quit:
 			return

